packages/database/src/migrations/20211125201506-SplitDataSourceTable-modifies-schema.js

```diff
@@ -0,0 +1,300 @@
+  await db.runSql(`
+    ALTER TABLE data_source
+    ADD COLUMN type data_source_type;
+
+    UPDATE data_source
+    SET type = 'dataElement';
+
+    ALTER TABLE data_source
+    ALTER COLUMN type SET NOT NULL;
+  `);
+  // Move data groups info back into data_source table
+  await db.runSql(`
+    INSERT INTO data_source (id, code, service_type, config, type)
+    SELECT generate_object_id(), code, service_type, config, 'dataGroup'
+    FROM data_group;
+  `);
+  // Update foreign keys back to data_source table
+  await updateForeignKey(
+    db,
+    'data_element_data_group',
+    'data_group_id',
+    'data_group',
+    'data_source',
+  );
+  await updateAndRenameForeignKey(
+    db,
+    'survey',
+    'data_group',
+    'data_group_id',
+    'data_source',
+    'data_source_id',
+  );
+  // Drop the data_group table
+  await db.runSql(`
+    DROP TABLE data_group;
+  `);
+
+  await updateAnalytics(db, 'data_source', 'data_element');
+};
+
+exports._meta = {
+  version: 1,
+};
+
+const updateAnalytics = async (db, newTableName, oldTableName) => {
+  await db.runSql(`
+    CREATE OR REPLACE FUNCTION build_analytics_table(force BOOLEAN default FALSE) RETURNS void AS $$
+    declare
+      tStartTime TIMESTAMP;
+      source_table TEXT;
+      source_tables_array TEXT[] := array['answer', 'survey_response', 'entity', 'survey', 'question', '${newTableName}'];
+      pSqlStatement TEXT := '
+        SELECT
+          entity.code as entity_code,
+          entity.name as entity_name,
+          question.code as data_element_code,
+          survey.code as data_group_code,
+          survey_response.id as event_id,
+          CASE 
+            WHEN question.type = ''Binary'' OR question.type = ''Checkbox'' 
+            THEN CASE 
+              WHEN answer.text = ''Yes'' THEN ''1'' 
+              ELSE ''0'' 
+            END 
+            ELSE answer.text
+          END as value,
+          question.type as type,
+          to_char(survey_response.data_time, ''YYYYMMDD'') as "day_period",
+          concat(
+            extract (isoyear from survey_response.data_time), 
+            ''W'', 
+            to_char(extract (week from survey_response.data_time), ''FM09'')) 
+          as "week_period",
+          to_char(survey_response.data_time, ''YYYYMM'') as "month_period",
+          to_char(survey_response.data_time, ''YYYY'') as "year_period",
+          survey_response.data_time as "date"
+        FROM
+          survey_response
+        INNER JOIN
+          answer ON answer.survey_response_id = survey_response.id
+        INNER JOIN
+          entity ON entity.id = survey_response.entity_id
+        INNER JOIN
+          survey ON survey.id = survey_response.survey_id
+        INNER JOIN
+          question ON question.id = answer.question_id
+        INNER JOIN
+          ${newTableName} ON ${newTableName}.id = question.${newTableName}_id
+        WHERE ${newTableName}.service_type = ''tupaia'' AND survey_response.outdated IS FALSE';
```

Comment:
Updating the analytics function means this may need to change depending on when the approval workflow work is merged, since that also updated this function

#3535